Adding functor input capability to ParsedAux

kyleeswanson opened the pull request:
## Reason
This PR supports functor input capability for ParsedAux class - contributes to Issue #21244 
<!--Why do you need this feature or what is the enhancement?-->




<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

Commit:
Added Functor Support to ParsedAux.C and ParsedAux.h, updated ParsedAux.md

Commit:
Adding test file for new ParsedAux functor capability

pbehne reviewed (changes requested):
Overall: great job!

While my requested changes may look like a lot, you have already done the most important work: adding new capability. My small suggestions are aimed at making sure that we are following MOOSE's [coding standards](https://mooseframework.inl.gov/sqa/framework_scs.html) and trying to make this Auxkernel idiot-proof at the input file level.

General comments:
1. I don't see a reference to the issue that this PR requests. I think it is related to #21244. If a related issue does not exist, you need to create one and reference it.
2. Please edit the description of the PR (you left it blank) to describe what the PR does. Reference the issue it addresses.
3. In order for your PR to merge, it needs to pass CIVET. One of the things that CIVET requires is that at least one of your commit messages should reference the relevant issue number (i.e., `Addressed code review suggestions related to issue #21244.`). Note the hashtag is crucial for CIVET to interpret the number as a reference.

Let me know if you have any questions or are ready for a re-review!

Commit:
Adding functor input capability to ParsedAux class - Issue #21244

Co-authored-by: Patrick Behne <31858053+pbehne@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2291866/) on 94b1894 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27959/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27959/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 705cbc287af66009b4e7cec4c8c17730387c976f`

kyleeswanson commented on framework/src/auxkernels/ParsedAux.C:
```diff
@@ -57,6 +67,14 @@ ParsedAux::ParsedAux(const InputParameters & parameters)
+  else
+    for (const auto & name : _functor_names)
+      variables += (variables.empty() ? "" : ",") + name;
+
```
Guillaume said don't worry about constant_names, constant_expressions, but I did update for if `_functor_symbols.size() != _functor_names.size()`

pbehne replied:
This is weird, I don't see a "resolve" button for this conversation.

pbehne replied:
I'll try deleting the whole comment.

GiudGiud commented on framework/doc/content/source/auxkernels/ParsedAux.md:
```diff
@@ -6,6 +6,8 @@ The parsed expression may contain:
 
 - variables (`coupled_variables` parameter)
 
+- functors (`functor_names` or `functor_symbols` parameter)
```
```suggestion
- functors (`functor_names` and/or `functor_symbols` parameter)
```

Commit:
Added error for size check between functor names and symbols - Issue #21244

Minor fixups - Issue #21244

GiudGiud committed:
Add checking for variable names vs functors
Issue #21244

Update framework/src/auxkernels/ParsedAux.C

Co-authored-by: Patrick Behne <31858053+pbehne@users.noreply.github.com>

Apply suggestions from code review

Co-authored-by: Patrick Behne <31858053+pbehne@users.noreply.github.com>

Fixed clang-tidy and formatting of ParsedAux.C
Issue #21244

Commit:
Add testing - Issue #21244

Added invalid expression test
Issue #21244

pbehne commented on framework/include/auxkernels/ParsedAux.h:
```diff
@@ -39,4 +39,16 @@ class ParsedAux : public AuxKernel, public FunctionParserUtils<false>
+  std::vector<std::string> _functor_symbols;
+
+  /// Vector of pointers to functors
+  std::vector<const Moose::Functor<Real> *> _functors;
```
Not making const because `_functors` cannot be populated in the initializer list.

pbehne commented on framework/src/auxkernels/ParsedAux.C:
```diff
@@ -48,17 +54,66 @@ ParsedAux::ParsedAux(const InputParameters & parameters)
+  else
+    for (const auto & name : _functor_names)
+    {
+      if (_use_xyzt && (std::find(xyzt.begin(), xyzt.end(), std::string(name)) != xyzt.end()))
```
```suggestion
      // Make sure symbol is not x, y, z, or t
      if (_use_xyzt && (std::find(xyzt.begin(), xyzt.end(), std::string(name)) != xyzt.end()))
```

pbehne reviewed (changes requested):
Just a few more nitpicky requests.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2298435/) on 8a0a5b4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27959/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27959/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d964da47455519058f746abe48a06dbb24590c62`

GiudGiud commented:
@pbehne

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2298640/) on ba44786 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27959/site)

This comment will be updated on new commits.

Commit:
Eliminated code duplication in sanity checks
Issue #21422

pbehne reviewed (approved):
Looks good! Great job, @kyleeswanson!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2301716/) on b8ac89d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>62dc98</th>
      <th colspan=4>#27959 b8ac89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.06%</td>
      <td>85.07%</td>
      <td>+0.00%</td>
      <td>96.15%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104412</td>
      <td>104454</td>
      <td>+42</td>
      <td>50</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18335</td>
      <td>18336</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27959/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27959/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27959/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Thanks for the review!
